File: recomp-mobile/app/src/main/java/com/refactor/app/api/MassUnit.kt
package com.refactor.app.api

/**
 * Converts between the user's chosen entry unit and the pounds every set log,
 * progression calculation, and personal record is stored in.
 *
 * Storage stays in pounds unconditionally — [Progression], `PersonalRecordStore`, and the
 * server all assume it. Only what the user types and reads is converted, so a metric
 * lifter's numbers stay comparable with their own history and with the web app.
 *
 * Mirrors iOS `MassUnit`.
 */
enum class MassUnit(val label: String) {
    POUNDS("lbs"),
    KILOGRAMS("kg");

    /** Convert a value the user typed into stored pounds. */
    fun toPounds(value: Double): Double = when (this) {
        POUNDS -> value
        KILOGRAMS -> value / LBS_TO_KG
    }

    /** Convert stored pounds into the unit the user reads. */
    fun fromPounds(pounds: Double): Double = when (this) {
        POUNDS -> pounds
        KILOGRAMS -> pounds * LBS_TO_KG
    }

    /** Display string with at most one decimal place, trimming a trailing `.0`. */
    fun display(pounds: Double): String = trimmed(fromPounds(pounds))

    companion object {
        const val LBS_TO_KG = 0.45359237

        /** `unitSystem` is the profile's raw string ("us" | "metric"). */
        fun forSystem(unitSystem: String?): MassUnit =
            if (unitSystem == "metric") KILOGRAMS else POUNDS

        fun trimmed(value: Double): String {
            val rounded = Math.round(value * 10.0) / 10.0
            return if (rounded % 1.0 == 0.0) rounded.toInt().toString()
            else String.format("%.1f", rounded)
        }
    }
}

File: RecompSwift/RefactorKit/Sources/Utilities/MassUnit.swift
import Foundation

/// Converts between the user's chosen entry unit and the pounds that every set log,
/// progression calculation, and personal record is stored in.
///
/// Storage stays in pounds unconditionally — `Progression`, `PersonalRecordStore` and
/// the server all assume it. Only what the user types and reads is converted, so a
/// metric lifter's numbers stay comparable with their own history and with the web app.
public enum MassUnit: String, Sendable {
    case pounds
    case kilograms

    public init(system: MeasurementSystem) {
        self = system == .metric ? .kilograms : .pounds
    }

    public var label: String {
        switch self {
        case .pounds: return "lbs"
        case .kilograms: return "kg"
        }
    }

    /// Convert a value the user typed into stored pounds.
    public func toPounds(_ value: Double) -> Double {
        switch self {
        case .pounds: return value
        case .kilograms: return value / WearableMassStoredPounds.lbsToKg
        }
    }

    /// Convert stored pounds into the unit the user reads.
    public func fromPounds(_ pounds: Double) -> Double {
        switch self {
        case .pounds: return pounds
        case .kilograms: return pounds * WearableMassStoredPounds.lbsToKg
        }
    }

    /// Display string with at most one decimal place, trimming a trailing `.0`.
    public func display(fromPounds pounds: Double) -> String {
        Self.trimmed(fromPounds(pounds))
    }

    public static func trimmed(_ value: Double) -> String {
        let rounded = (value * 10).rounded() / 10
        return rounded.truncatingRemainder(dividingBy: 1) == 0
            ? String(Int(rounded))
            : String(format: "%.1f", rounded)
    }
}
